### ISSUE TITLE
Story for pull requests

### ISSUE THREAD

**Issue by aryairani** [unisonweb/unison#867]
#### Proposal summary
* [x] add `pr.create <base-path> <head-path>` command: generates a message providing the `pr.load` command line, and a diff summary. (#966)
  * [x] include added/deleted/modified patches in diff summary (#965)
  * [x] pr.load should cause only one root update #1201 
  * [ ] pr.create shouldn't produce useless numbering #1203 
  * [x] pr.load should take optional destination path #1204
* [x] push/pull should transmit even unnamed dependencies (#930) (not applicable until #940?)
* [x] push/pull should support specifying a remote path (#931)
* [x] add `pr.load <base-path> <head-path>` command: forks these two revisions into subpaths, also does a preliminary merge into a subpath ~, lists todos~
  * [x] ~write a parser for unison short urls~ (implemented but then deleted in 8d7da9ef7d6dec256e67b7899eba542dc205a238)
  * [x] update git support to be able to push/pull from a particular remote Unison namespace
* add "diff" commands suitable for code review
  * [x] ~dump out old and new branch files, for external diff tool~
  * [ ] `diff.term <old> <new>` (#1215)
    * possibly/eventually some heuristic to decide when expressions have changed, vs moved or pushed into a subexpression
    * eventually identifying authorship of subexpressions
    * we might want eventually to design a custom pretty-printing algorithm for diffs; perhaps wrapping lines differently at diff boundaries
  * [ ] `diff.type <old> <new>` (#1216)
  * [x] `diff.namespace <old> <new>`
  * [ ] `diff.patch <old> <new>` (#1217)
  * [ ] `diff.namespace.verbose <old> <new>`, which aggregates all these diffs 
    * into some directly editable and parseable form?  (#1218)
* [x] ~link previous definition to new definitions on `update`~ where we left this last was that we didn't want to dynamically generate metadata during the propagation step.  but why?
* [ ] possibly author Documentation objects, which can be exchanged during the review process (#1219)

***

## Code Review and Pull Requests

The code review process involves supporting a few major pieces:

1. The contributor publishes the code to be reviewed to some git repo.
2. The contributor sends a pull request to the reviewer.
3. The reviewer retrieves and reviews the proposed changes

   * Comparing changed branch contents
   * Comparing old/new definitions of replaced names
4. Reviewer discusses / requests changes, or merges the PR.
5. Attribution

This doc walks through how these steps are implemented in Github, git, and give some proposals for ucm.

### Step 1: Contributor publishes code

Obvious / somehow / whatever.

### Step 2: Contributor sends a pull request to the reviewer

A pull request may include:

1. a base repository, where the maintained code is published
2. a base branch known to the reviewer, to compare against
3. the head repository, where the proposed code is published
4. the proposed branch, published in the head repository.
5. a motivating textual description of the changes

**<u>Github</u>**: 

On Github, there is a nice GUI for constructing and submitting PRs.  The two repositories are presumed to exist on Github, and the first four bits of information are filled in through auto-completing drop-downs. 

<img src="https://user-images.githubusercontent.com/538571/67627604-3fda5200-f82d-11e9-92df-0722b7377353.png" style="zoom:25%;" />

The contributor motivates the proposed changes in a text field, which can link to or transclude code from other Github entities.

**<u>git</u>**:

Command-line git will help generate a pull request via the (`git request-pull <start> <url> [<end>]`)[https://git-scm.com/docs/git-request-pull] command.  It takes three arguments (the base revision, the head repo, and the proposed revision).  The base repo is indicated out of band, though if the base revision is specified in the form of a hash, it does narrow the choices.  

Example:

```
$ git push git@github.com:noahhaasis/unison.git todo-builtin
$ git request-pull master git@github.com:noahhaasis/unison.git todo-builtin
```
produces the following message; `git request-pull -p` would include textual diff output as well.
```
The following changes since commit 6bce3188e15385336d2c115ac9d4a0488dc971cd:

  Merge pull request #875 from noahhaasis/guard-and-or (2019-10-19 13:06:19 -0400)

are available in the Git repository at:

  git@github.com:noahhaasis/unison.git todo-builtin

for you to fetch changes up to 07f5a83944f96e470859378af45db251632a6083:

  Add `error` and `todo` builtin (2019-10-21 14:46:26 +0000)

----------------------------------------------------------------
Noah Haasis (1):
      Add `error` and `todo` builtin

 parser-typechecker/src/Unison/Builtin.hs     | 2 ++
 parser-typechecker/src/Unison/Runtime/Rt1.hs | 1 +
 2 files changed, 3 insertions(+)
```

The generated output includes the commit message for `<end>`, the url and and a diff from `<start>` to `<end>`.  I'm not sure whether or not it checks that `<start>` and `<end>` exist at `<url>`, but it may.

The generated pull request and motivating description is typically emailed to a reviewer or mailing list.

Notes:

* It looks like `git request-pull` doesn't necessarily indicate which project or repo the proposed changes are being proposed to, though the commit indicated by `<start>` may imply it.

<u>**ucm proposal**</u>:

I don't know.  We could add a ucm command "pr.create" to generate a message like `git request-pull` does, and the contributor could send it to the maintainer via Github issues, or via email or IM.  Or we could simply propose a textual template that contributors populate themselves.

It would be nice if PRs to public projects were done in some way we could easily count, for usage metrics, but this might be better handled with a more comprehensive, opt-in ucm usage tracking.

### 3. The reviewer retrieves and reviews the proposed changes

**<u>git</u>**

From the out-of-band PR, you infer the target codebase, and learn the URL and commit from which to retrieve the proposed commit.  The out-of-band PR may include a text diff of the contributor's changes since their base version, if they specified the `-p` option.  

You may use `git diff` locally to compare against an updated base commit; build & test the application, etc.  Any feedback about the commit is also provided out-of-band (email, discussion forum, etc.)

**<u>Github</u>** 

Github already has all the source code, so there's nothing to retrieve unless you need to execute it locally or something.  It's got a lovely UI for navigating between changes to the textual diff (sad-face).  Apparently you can directly write changes to the repo/branch associated with the contributor's PR.  

One or more reviewers can leave feedback in a variety of places:

* associated with the PR, as part of the associated thread
* associated with individual commits, 
* associated with lines of text files within individual commits, 
* associated with the PR or individual lines of the branch head, as part of the new "Review" process

Review text which references other Github issues is automatically hyperlinked; text which references Github source line ranges is (now) automatically transcluded into the rendered comments.

**<u>ucm proposal</u>**

Optionally introduce a `pr.load <base> <head>` command, where `<base>` and `<head>` each represent published unison paths; optionally including git url, git commit/branch, Unison namespace hash, Unison namespace path.  

> **<u>Sidebar on Unison URLs:</u>**
>
> Frustratingly, [git branches](https://wincent.com/wiki/Legal_Git_branch_names) may include `/`, `.`, `#`, `=`, etc. which make them hard or impossible to parse a branch name among `/`-separated elements (as is needed to parse branches out of Github urls) without using IO to check whether a given parse corresponds to an actual branch.  At least `:` isn't allowed in a branch name, though it is unfortunately allowed in scp-style git repo addresses (_e.g._ `git@github.com:foo/bar`).  Saving grace: Github doesn't allow `#` or `:` or `/` in repo names, so that's something.
>
> `gh:<user>/<repo>[/<treeish>][:[<nshash>[.<path>] | [.]<path>]]`
>
> Examples:
>
> ```
> gh:aryairani/unisonbase
> gh:aryairani/unisonbase:libs.v1
> gh:aryairani/unisonbase:#abc123sbh
> gh:aryairani/unisonbase/mybranch
> gh:aryairani/unisonbase:#abc123sbh.path
> gh:aryairani/unisonbase/dev
> gh:aryairani/unisonbase/dev:mybranch
> ```
>
> There's a question of whether we want to embed the url type ("namespace") as well, to make sure we know how we can easily distinguish other kinds of things we might want to refer to by url:
>
> ```
> gh:<user>/<repo>[/<treeish>]:namespace:[<nshash>.<path>|<path>]]
> gh:<user>/<repo>[/<treeish>]:term:[<displayname>#]<hash>
> gh:<user>/<repo>[/<treeish>]:type:[<displayname>#]<hash>
> ```
>
> For @pchiusano: This "short" URL scheme that we define can be embedded in a real URL for whatever later purposes, e.g. https://unisonweb.org/code-browser#gh:aryairani/unisonbase:#abc123sbh.path

At any rate, this proposed `pr.load` command would work something like this:

```
.pulls.base.contributor1> pr.load gh:user/topicrepo:path gh:org/baserepo:path2

  I forked contributor/repo/path into `topic` (name tbd)
  And forked destination/repo2/path2 into `target` (name tbd)

  They've been auto-merged into `merged`, with 3 todos.

  Use `diff target merged` to see what has been updated
  Use `todo merged merged.patch` to see what work is remaining in for the merge
  Use `push merged destination/repo2/path2` to push the changes.

.pulls.base.contributor1>
```

`load-pr` is really just a convention-following mix of `pull` and `merge`, so it's a low priority.  (Maybe even implemented as a `{CodebaseManager}`function.)

We would introduce a few diff commands too:

`diff.term x1 x2`: text-based or ABT-based diff of terms.

`diff.type x1 x2`: I don't even know for this one?  Is a text-based diff even useful here?

`diff.namespace x1 x2`: lists added, deleted, and modified definitions & patches.

`diff.patch x1 x2`: lists added, deleted, and modified entries.

`diff.namespace.verbose x1 x2`: pretty-print added & deleted defs & patches, pretty-diff modified defs & patches.  Great if the output could also be a parseable .u file: like `diff.edit x1 x2`

### Step 4: Commit or return for revisions

**<u>Github</u>**

Discussion & change requests happen on the Github website, with notifications sent to involved parties.  When the reviewer is finished, they can "Merge pull request" or wait for revisions.

**<u>git</u>**

`git push` when you're ready; any revisions are done out-of-band.

**<u>ucm proposal</u>**

Comments and revisions could be done out-of-band.  The maintainer can merge and publish changes like normal, when ready.

### Step 5: Attribution

Github and git attribute individual commits to individual commiters, by name and email address.  Both tools can use heuristics to assign blame or praise to individual committers by line, in a text file.

Github creates graphs and tables of contributors at https://github.com/unisonweb/unison/graphs/contributors.

Unison doesn't currently have any facility for tracking contributor identity, but it's a valuable thing to many in the otherwise uncompensated field of open-source software development.

**<u>ucm proposal</u>**

* We prompt the user at ucm startup to populate `.unisonConfig` with their identity, and attach it to `add`ed/`update`d definitions as metadata.
* We could use `hub` instead of `git`, and use `hub api user` to determine their github id, and use that as metadata.  It would require an OAuth authentication step, handled by `hub`.  I'm thinking this doesn't buy us anything.

* Simply adding author metadata to definitions doesn't immediately give you the whole story on the full collaborative authorship of a definition.  I'd say we don't have an accurate way to ask, "Who wrote the previous version of this definition?" unless we link the previous version of the definition as metadata to the updated version, which we could do.  Using the definition's name as a proxy for this info is no good IMO.  We could use this edit chain eventually do some syntax highlighting showing who's written the surviving sub-elements of the definiton.

**Comment by atacratic** [comment]
Cool, happy to see this 😀

- I'm thinking people may still want to create actual GitHub PRs, even though the GitHub diffs are no use, to get the notifications band workflow stuff. So they'd paste the ucm pull command text into that PR.

- Do you have plans for showing diffs? Just comparing old and new definitions by eye with no diff cues (green for additions etc) isn't going to fly in the long term. Are you thinking that viewing branch diffs is a job for the browser based code viewer one day? What about at the ucm prompt? I don't think diff highlighting necessarily implies a tree/structure aware diff algorithm. You could start with doing a text based diff and then feeding the results back as annotations into the original tree.

- would review comments ideally be metadata? So people can work through them, maybe with the help of the to-do command? 

- can the reviewer suggest code changes? Where do they get pushed to? Easy for author to pull those back in?

- can we make the reviewer's interface look like a kind of scratch file? Output changes and diffs, wait for the user to annotate it with their comments, and slurp back the result? Otherwise we might be waiting a long time for a nice github-like reviewer experience. Maybe the scratch file could contain the new code in parseable format, and the diff cues are done as snippets of old code in comments. To make it easy to slurp in suggested code changes from the reviewer, and for the reviewer to actually understand the new code.

- what happens on round 2 when the reviewer is pulling in a branch they already pulled in a previous version of before?

- I think we do need some kind of bulk text export of the changes. The reviewer doesn't want to type out 100 different view commands to see all the new functions.

**Comment by aryairani** [comment]
Thanks @atacratic, and yeah my initial post was intentionally vague on these points, just to get out a starting point.

> * I'm thinking people may still want to create actual GitHub PRs

Honestly I think the Github PR interface is going to be completely useless for Unison.  Down the line, I would like to see a web-based tool for doing code reviews, with all of the features of Github PRs and probably more, but it's not feasible in the short term.  I would be happy to review a mockup of either, though; but for the time being, this is not part of the plan.

> * Do you have plans for showing diffs?

Yes, definitely! I was thinking a highlighted, whitespace-aware, text-based diff like what I get in IntelliJ would be good.  Although, after a moment's further reflection, I think text-based diff is a waste of time, compared to a "good" tree/structure aware diff algorithm.  So the question is, is there a totally ready off-the-shelf text diff implementation somewhere?

> and then feeding the results back as annotations into the original tree.

what?

> * would review comments ideally be metadata? So people can work through them, maybe with the help of the todo command?

Yeah, it would be nice if they were stored in the codebase somehow; maybe part of an evolving PR definition, or maybe metadata attached to an immutable PR?  Initially, however, they could be 100% out-of-band.

> * can the reviewer suggest code changes? Where do they get pushed to? Easy for author to pull those back in?

One option is that the reviewer can suggest code changes out of band.  Another option is that the reviewer could submit a PR back to the contributor, reusing these same mechanisms.

> * can we make the reviewer's interface look like a kind of scratch file? Output changes and diffs, wait for the user to annotate it with their comments, and slurp back the result? Otherwise we might be waiting a long time for a nice github-like reviewer experience. Maybe the scratch file could contain the new code in parseable format, and the diff cues are done as snippets of old code in comments. To make it easy to slurp in suggested code changes from the reviewer, and for the reviewer to actually understand the new code.

May...be...?  The scratch file isn't going to be colorized though.

> * what happens on round 2 when the reviewer is pulling in a branch they already pulled in a previous version of before?

Not sure I'm answering the right question, but I would imagine the reviewer could choose to FF merge it onto the previous temporary pull, or into a new namespace.

> * I think we do need some kind of bulk text export of the changes. The reviewer doesn't want to type out 100 different view commands to see all the new functions.

I was thinking we could use numbered command substitution and have the PR version of `todo`.

It would be nice if `ucm` helped you keep track of "garbage" like old or incomplete PRs.  I think it's nice that in Github, PRs are separate from the code.  I'm not totally comfortable with the story for metadata in general (including patches) yet.  It would be nice if `ucm` helped you keep track of incompletely applied patches too, but it's hard to identify them since you can't read the user's mind.  Maybe stronger "convention over configuration" is the answer for this, but I'm not sure what they should be.

**Comment by atacratic** [comment]
### Viewing diffs

I think having to type a command for every function change you want to look at is a pain.  If I'm reviewing some code, the maximum I want to have to do to read through it is scroll my mouse wheel.

How about a command, something like `review some-branch`, or `review some-branch after`, or `review old-branch new-branch`, that spits out a scratch file `new-branch.u` looking like the following?

```haskell

-- added definition foo
foo : Nat -> Nat
foo n = n + 1

-- removed definition bar
-- bar#6y241 : Nat -> Boolean

-- changed definition baz
-- new definition:
baz : '{IO} ()
baz _ = printLine "Hello"
        --<++
        printLine "cruel"
        --++>
        printLine "world!"

```

We could have heuristics for a nice order to output the changes (dependency graph leaves first, types first...)

Note `baz` in the example above is output with inline textual diff cues, which is assuming a tree-aware diff.  

We could skip implementing that diffing logic, especially if we can also output the old version.  So `review some-branch before` or `review.before old-branch new-branch` spitting out a file `old-branch.u` like...

```haskell

-- definition bar
bar : Nat -> Boolean
bar n = case n of 
  0 -> true
  1 -> false
  _ -> bar (drop (drop n))

-- definition baz
-- old definition:
baz : '{IO} ()
baz _ = printLine "Hello"
        printLine "world!"
```

Notice how this is set up so that the user can diff the two scratch files themselves in their preferred tool and get nice results.  Maybe we can have a command to write the two files and call that tool directly, and config for what tool to call out to (or use git's `diff.tool` value).  (Maybe another command to merge two versions of a definition into a new scratch file using an external merge tool.)

Why did you get to thinking that text based diff would be a waste of time?

#### Where do PRs live?

Are you thinking that 'PR' is a concept that `ucm` is aware of?  I hadn't gotten that far.  I was thinking that `PR` is only a concept that makes sense when it lives on a webserver, because (a) it needs a URL, and (b) it needs a user database, notification system, and conversation page.  

I was thinking that `ucm` knows about branches, and it maybe knows about open review comments as metadata.

In the longer term of course we want github to support unison repos properly :-)

And in the medium term maybe there's a unison-specific web-based front-end to github, for viewing diffs and comments.  

But until then, for open source unison code, I'm thinking that people store it in github regardless, and communicate with each other about desired changes using github PRs - if nothing else because github has a verified user database and people can communicate without exchanging email addresses.

So even though the 'files changed' tab of a PR is useless, people ignore it, and instead run a ucm command, or visit a code browser URL, and those things pull the branch from github and present it properly.  

Not sure I've written the above very well - maybe we can slack about it?

#### Comments

Agree that comments can initially be out of band.  People can annotate the `new-branch.u` file and send it to each other.  (Might be worth touching on this point again as part of the design of the API comments mechanism - might be easy to add another kind of comment into the mix.)

**Comment by aryairani** [comment]
> ### Viewing diffs
> I think having to type a command for every function change you want to look at is a pain. If I'm reviewing some code, the maximum I want to have to do to read through it is scroll my mouse wheel.

Mm yes, I'm inclined to agree.

> How about a command, something like `review some-branch`, or `review some-branch after`, or `review old-branch new-branch`, that spits out a scratch file `new-branch.u` looking like the following?

Something like this could be good.

> Note `baz` in the example above is output with inline textual diff cues, which is assuming a tree-aware diff.

Is there an example that really highlights a tree-aware diff?  This one doesn't seem to.  But it looks like a reasonable text diff.

> Why did you get to thinking that text based diff would be a waste of time?

I was just thinking of changes a minor change (e.g. something necessitating a change in indentation) causes a huge text diff, where the interesting changes are lost in the noise.

> Are you thinking that 'PR' is a concept that `ucm` is aware of? I hadn't gotten that far. I was thinking that `PR` is only a concept that makes sense when it lives on a webserver, because (a) it needs a URL, and (b) it needs a user database, notification system, and conversation page.

There are some projects (e.g. git, linux) handle pull requests by email (and will likely continue to do so as long as Linus runs them), with no separate user database, notification system, or conversation page. They send an auto-generated email that describes how to check out the proposed code, use regular `git` tools to compare commits, and use email to discuss requested changes.  I was planning to model ucm's pull requests similarly to start.

https://git-scm.com/docs/git-request-pull
https://hackernoon.com/how-to-git-pr-from-the-command-line-a5b204a57ab1

> I was thinking that `ucm` knows about branches, and it maybe knows about open review comments as metadata.

Yeah I'm a bit fuzzy on this, but it's an option.  The immediate goal is to figure out what's the minimum needed in terms of new features in order to have a passable experience doing code reviews, so people can collaborate on libraries.  Then definitely a roadmap to luxurious code reviews :)

> And in the medium term maybe there's a unison-specific web-based front-end to github, for viewing diffs and comments.

Yeah something like this could be good.  Team `unisonweb/elm-browser` ran into a github API throttling that we'll have to work around somehow.

> So even though the 'files changed' tab of a PR is useless, people ignore it, and instead run a ucm command, or visit a code browser URL, and those things pull the branch from github and present it properly.

Yeah... maybe?

> Not sure I've written the above very well - maybe we can slack about it?

Sure :)

> #### Comments
> Agree that comments can initially be out of band. People can annotate the `new-branch.u` file and send it to each other. (Might be worth touching on this point again as part of the design of the API comments mechanism - might be easy to add another kind of comment into the mix.)

**Comment by atacratic** [comment]
> > Note baz in the example above is output with inline textual diff cues, which is assuming a tree-aware diff.

> Is there an example that really highlights a tree-aware diff? 

Ah yeah, you're right, you could do this without tree diffing.

> a minor change (e.g. something necessitating a change in indentation) causes a huge text diff, where the interesting changes are lost in the noise

People are used to this in their tooling today, so it's tolerable.  And most diff tools make indentation changes look subtle so not too noisy.

Note that if we're outputting review scratch files, we can make both the before and after files use the new names for stuff, which cuts out noise.

before file:
```haskell
-- definition hoop - renamed hoopy
hoopy : Nat -> Nat

-- definition foo
-- old definition
foo n = 2 * (hoopy n)
```
after file:
```haskell
-- definition hoop - renamed hoopy
hoopy : Nat -> Nat

-- definition foo
-- new definition
foo n = 3 * (hoopy n)
```
> There are some projects (e.g. git, linux) handle pull requests by email 

Speaking as an individual potential Unison user now, rather than a small-time contributor: I cannot see myself ever collaborating with a bunch of people on an open source library by email.  It's a massive regression from having the code, and conversations about possible changes to the code, hosted on a public server that acts as a message board.  If the Unison docs say 'send this text to your collaborators out-of-band', then I think I will be sending it to them by pasting it into a github issue or PR!

Basically I think you have to define 'out of band' - how do you actually see this working for people in the first weeks after this function lands?  Where are they publishing their code?  What public places are they publishing their email addresses?  (I use github's facility to keep mine hidden.)  Are they all making first rendezvous via the unison slack or what?

**Comment by atacratic** [comment]
You said via slack that the problem with GitHub PRs is that a raw git merge merges at the root namespace, which is not generally what we want.  Could you say a bit more about that? And do you think people can still use GitHub for hosting unison libraries?

Idea 1: could ucm push to a (git) **branch** in the remote repo, in a namespace-aware way? So not pushing all files but just ones included in the namespace being pushed?

Idea 2: still use GitHub for code hosting, but recommend a workflow in which people raise issues instead of PRs, with the issue text including ucm commands for reviewing/pulling.

**Comment by aryairani** [comment]
Idea 2 could work, we would be treating issues as email, which is basically what they already are; but a unison PR doesn’t correspond to a git/github PR.

I believe we can use a process that’s analogous-but-not-identical to the git command-line PR, and perhaps someday also have one that’s analogous to the github PR, but it doesn’t fit the model because it’s not really a git PR.

I guess you could have imagine a git PR where I say “here’s what your final repo should look like after merging my stuff” but that at the moment that feels like it assumes too much about the contributor’s knowledge of the reviewer’s codebase, which could be organized arbitrarily.  Or maybe it means you’d need a git repo for any point of the tree you’d like to collaborate on; this seems inflexible too. Will think about it.

**Comment by aryairani** [comment]
Some notes from discussion today:
`load-pr` (name tbd) takes two args: (contributor repo + path) and (destination repo + path)
1. PR command is sent to reviewer
2. Reviewer creates namespace for reviewing the merge
```
.pulls.base.contributor1> load-pr contributor/repo/path destination/repo2/path2

  Forked contributor/repo/path into `topic` (name tbd)
  Forked destination/repo2/path2 into `target` (name tbd)
  Merged into `merged`

  Use `diff target merged` to see what has been updated
  Use `todo merged merged.patch` to see what work is remaining in for the merge
  Use `push merged destination/repo2/path2` to push the changes.
```

Need to flesh out more about actually diffing though.

**Comment by atacratic** [comment]
@aryairani I took a look at your edited version of the issue text, looks promising!

You've listed the `diff.namespace.verbose <old> <new>` possibility, cool :smile:  I was also suggesting a variant that outputs the *previous* code, so that ucm could call out to a graphical diff tool to do the actual diff and presentation logic.  I did an actual mock-up to try and show it better (this is a screenshot of meld):

![image](https://user-images.githubusercontent.com/21991324/67641814-2a345d80-f8fd-11e9-8f23-4151663e976a.png)

Do we actually need all the other diff commands, for now at least, if we have this?

.

Also, separately, a minor thing on the following:
> We prompt the user at ucm startup to populate .unisonConfig with their identity

- Prompting on first need would be nicer than prompting on first start, I think.
- Did you consider reading git's username/email config?  Since we're depending on git already, and when we call out to git to push and pull, it's writing its configured username/email out to the remote repos in its commit metadata?

**Comment by aryairani** [comment]
> Did you consider reading git's username/email config? 
👍

**Comment by atacratic** [comment]
@aryairani  After our chat on slack, here's another proposal for these commands, for you to throw into the mix as you work out what to do.  

#### Commands summary


These two just as you have them above:
```
pr.load <head-path> <base-path>
pr.create <head-path> <base-path>
```

Instead of the various diff commands listed above, how about instead the following:

```
namespace.diff-summary <from> <to>
`namespace.diff-summary .foo1 .foo2`      summarizes the changes between .foo1 and .foo2

namespace.diff <from> <to>
`namespace.diff .foo1 .foo2`              shows the changes between .foo1 and .foo2 using 
                                          `git difftool`

namespace.diff-edit <from> <to> [<to-file>] [<from-file>]	
`namespace.diff-edit .foo1 .foo2`         prepends any definitions from .foo2 which 
                                          differ from the corresponding definitions 
                                          in .foo1, to the most recently edited 
                                          scratch file
`namespace.diff-edit .foo1 .foo2 after.u` prepends any definitions from .foo2 which 
                                          differ from the corresponding definitions 
                                          in .foo1, to after.u
`namespace.diff-edit .foo1 .foo2 after.u before.u` 
                                          prepends any definitions from .foo2 which 
                                          differ from the corresponding definitions 
                                          in .foo1, to after.u and before.u respectively

```

#### Transcript - loading a PR and viewing in a graphical diff tool

Load the PR...

```
.pulls.base.contributor1> pr.load gh:user/topicrepo:path gh:org/baserepo:path2

  I forked user/topicrepo/path into `topic` (name tbd)
  And forked org/baserepo/path2 into `target` (name tbd)

  They've been auto-merged into `merged`, with 3 todos.

  Use `namespace.diff-summary target merged` to see what has been updated
  Use `namespace.diff target merged` to see the updates in detail using `git difftool`
  Use `todo merged merged.patch` to see what work is remaining in for the merge
  Use `push merged destination/repo2/path2` to push the changes.

.pulls.base.contributor1>
```

View a summary of the changes...

```
.pulls.base.contributor1> namespace.diff-summary target merged

  + Adds / updates: 

    foo, baz

  - Deletes: 

    bar
```

View in a graphical diff tool...

```
.pulls.base.contributor1> namespace.diff target merged

⚙  git difftool /tmp/Unison-hd8261j/target /tmp/Unison-hd8261j/merged
```

![image](https://user-images.githubusercontent.com/21991324/67641814-2a345d80-f8fd-11e9-8f23-4151663e976a.png)

#### Transcript - loading a PR and viewing in a scratch file

Load the PR... (same as above)

View in a scratch file...

```
.pulls.base.contributor1> namespace.diff-edit target merged merged.u

  Scanned for differing terms, prepended the version from merged to merged.u.
```

Creates file merged.u...

```
-- removed definition bar#6y241
-- bar : Nat -> Boolean
-- bar n = case n of 
--   0 -> true
--   1 -> false
--   _ -> bar (drop (drop n))

-- definition baz
baz : '{IO} ()
baz _ = printLine "Hello"
        printLine "cruel"
        printLine "world!"

-- added definition foo
foo : Nat -> Nat
foo n = n + 1
```

This is just the right hand side of the graphical diff set up by namespace.diff.  

Unison will watch for changes in merged.u just as with any scratch file.  That provides a convenient way for the user to tweak the merged version of the code (and, in future, add comments to be sent back to the PR author.)

#### Transcript - setting up a graphical diff, controlling the file locations

Load the PR... (same as above)

Create a pair of scratch files...

```
.pulls.base.contributor1> namespace.diff-edit target merged merged.u base.u

  Scanned for differing terms:
   - prepended the version from target to base.u
   - prepended the version from merged to merged.u

  You can compare these by running the following command from the shell.
    git difftool base.u merged.u

```

Creates files merged.u and base.u (the same contents as would be used for the graphical diff by namespace.diff.)

#### Transcript - first use of namespace.diff (setup)

What they'll see first time depends on their existing git diff.tool config.  If they've already configured a tool they like then no problems, see above.

If there's no tool configured it might be something like...

```
.pulls.base.contributor1> namespace.diff target merged

⚠️ You may not have a diff tool configured.  
   Tip: consider installing a tool like meld (https://meldmerge.org/)
   Tip: `git config --global diff.tool meld`

⚙  git difftool /tmp/Unison-hd8261j/target /tmp/Unison-hd8261j/merged

This message is displayed because 'diff.tool' is not configured.
See 'git difftool --tool-help' or 'git help config' for more details.
'git difftool' will now attempt to use one of the following tools:
meld opendiff kdiff3 tkdiff xxdiff kompare gvimdiff diffuse diffmerge ecmerge p4merge araxis bc codecompare emerge vimdiff

Viewing (1/1): 'before.u'
fatal: external diff died, stopping at before.u    
```

Or it may be configured just to a text diff:

```
.pulls.base.contributor1> namespace.diff target merged

⚠️ You may not have a diff tool configured.  
   Tip: consider installing a tool like meld (https://meldmerge.org/)
   Tip: `git config --global diff.tool meld`

⚙  git difftool /tmp/Unison-hd8261j/target /tmp/Unison-hd8261j/merged

diff --git a/../before.u b/../after.u
index 32bc4d81..59e6e247 100644
--- a/../before.u
+++ b/../after.u
@@ -1,11 +1,16 @@
--- definition bar#6y241
-bar : Nat -> Boolean
-bar n = case n of 
-  0 -> true
-  1 -> false
-  _ -> bar (drop (drop n))
+-- removed definition bar#6y241
+-- bar : Nat -> Boolean
+-- bar n = case n of 
+--   0 -> true
+--   1 -> false
+--   _ -> bar (drop (drop n))
 
 -- definition baz
 baz : '{IO} ()
 baz _ = printLine "Hello"
-        printLine "world!"
+        printLine "cruel"
+        printLine "world!"
+
+-- added definition foo
+foo : Nat -> Nat
+foo n = n + 1
```

There'll probably be some experimentation needed to pop up the right warning in the right situations.

**Comment by aryairani** [comment]
Notes from conversation with @pchiusano and @runarorama 
* good to leverage already-written code where possible (e.g. external diff tools)
* maybe a numbered namespace diff and the ability to zoom to individual entries is enough
* want to help user understand term changes: what was modified as part of an auto-propagated dependency change vs what was modified manually by a human
* which dependency names should be used when displaying old / new terms for diffs?

**Comment by aryairani** [comment]
Github has this nice new system for keeping track of the changes you've already reviewed:
![image](https://user-images.githubusercontent.com/538571/67797108-96c36f80-fa57-11e9-9b34-563519cd84d7.png)

**Comment by aryairani** [comment]
* Refactored Git.hs to use `Maybe` for the git branch, and to use `-C dir` instead of `setWorkingDirectory` (implemented as `gitIn` and `gitTextIn`); hopefully breaking nothing in the process.

**Comment by aryairani** [comment]
Consolidate #796

**Comment by pchiusano** [comment]
@aryairani assuming you captured everything still outstanding as separate tickets, can you close this ticket?

**Comment by aryairani** [comment]
I didn't capture everything from the comments, but I captured everything from the summary.
